### PR TITLE
Clear conntrack entries for externalIP and LoadBalancer IP

### LINKED ISSUE #65228: Delete stale UDP conntrack entries with external ip usage
**Is this a BUG REPORT or FEATURE REQUEST?**:
 /kind bug

**What happened**:
Fix #59286 doesn't remove the stale udp conntrack session when an external ip is involved. The stale session lingers and continues to misdirect traffic to endpoints that are removed and it prevents sessions to new endpoints from being created.

**What you expected to happen**:
Fix #59286 removes stale udp conntrack sessions by cluster ip as its solution but if using external ip (ex. vip) the stale udp conntrack session can't be matched by cluster ip so it doesn't get removed. In this scenario, the stale session's fields contain the external ip, endpoint ip and port but not cluster ip.

**How to reproduce it (as minimally and precisely as possible)**:
With all inbound traffic routed through a vip, the following sflow-app has its traffic forwarded from an external ip 10.0.22.190 to its endpoint 10.32.0.83:6343:

NAME                      TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                       AGE
sflow-app                ClusterIP   10.105.23.70     10.0.22.190   6343/UDP                      6d

NAME                      ENDPOINTS                                          AGE
sflow-app                10.32.0.83:6343                                    6d

The conntrack session for the above service looks like this (notice the cluster ip of 10.105.23.70 is not part of the session):
udp      17 29 src=10.0.22.3 dst=10.0.22.190 sport=50401 dport=6343 [UNREPLIED] src=10.32.0.83 dst=10.32.0.1 sport=6343 dport=50401 mark=0 use=1

When the pod/endpoint is removed or it gets a new ip address, the conntrack session doesn't get removed by external ip or endpoints so it just lingers.

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-21T09:17:39Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

- OS (e.g. from /etc/os-release):
NAME="Ubuntu"
VERSION="18.04 LTS (Bionic Beaver)"

- Kernel (e.g. `uname -a`):
Linux firescope1 4.15.0-22-generic #24-Ubuntu SMP Wed May 16 12:15:17 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

/sig network

### PR DESCRIPTION
**Pull request by prameshj**
When an endpoint is deleted, the conntrack entries are cleared for
clusterIP but not for externalIP of the service. This change adds
that step.

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
When an endpoint is deleted, the conntrack entries are cleared for clusterIP but not for externalIP of the service spec and LoadBalancer ip from service status. This PR adds that step.
**Which issue(s) this PR fixes**:

Fixes #65228

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by steven-sheehy** [comment]
Thanks @prameshj for the fix. I can try to test locally if you have a docker image pushed somewhere.

Also, the fixes statement in the description has an extra slash, so referenced ticket won't get closed automatically.

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -613,6 +613,12 @@ func (proxier *Proxier) deleteEndpointConnections(connectionMap []proxy.ServiceE
 			if err != nil {
 				klog.Errorf("Failed to delete %s endpoint connections, error: %v", epSvcPair.ServicePortName.String(), err)
 			}
+			for _, extIP := range svcInfo.ExternalIPStrings() {
+				err := conntrack.ClearEntriesForNAT(proxier.exec, extIP, endpointIP, v1.ProtocolUDP)
```
Comment (hiep-huynh) [inline pkg/proxy/iptables/proxier.go]: This has the potential to indiscriminantly remove all sessions matching external IP to endpointIP regardless if it's stale or not.  Can this be written to be more selective especially if we know the port(s) involved in the stale session?

**Review by hiep-huynh — CHANGES_REQUESTED** [review]

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (prameshj) [reply pkg/proxy/iptables/proxier.go]: There is a check above(L608) to delete stale connections based on the NodePort. If user is not using NodePort and has specified TargetPort by name, that could have a different value for each endPoint. ServiceEndpoint currently does not expose this value.

**Review by hiep-huynh — APPROVED** [review]
I repeatedly tested our apps using latest test build with your fix. I deleted the pods with UDP services in order to reproduce/cause stale sessions and verified that the stale sessions were removed immediately. I also verified that new UDP sessions and traffic were able to be created to the new pod instances.

In doing so I realized that my concerns about inadvertent removal of sessions for matching extIP, endpointIP, and of UDP type doesn't apply for the reason that pod instance endpointIP are unique and in deleting the instance, it's desirable to clear all sessions related to this endpointIP to prevent routing black holes. I'm able to verify that your code accomplishes this by clearing sessions from all possible extIPs to the removed pod instance endpointIP.

**Comment by prameshj** [comment]
> I repeatedly tested our apps using latest test build with your fix. I deleted the pods with UDP services in order to reproduce/cause stale sessions and verified that the stale sessions were removed immediately. I also verified that new UDP sessions and traffic were able to be created to the new pod instances.
> 
> In doing so I realized that my concerns about inadvertent removal of sessions for matching extIP, endpointIP, and of UDP type doesn't apply for the reason that pod instance endpointIP are unique and in deleting the instance, it's desirable to clear all sessions related to this endpointIP to prevent routing black holes. I'm able to verify that your code accomplishes this by clearing sessions from all possible extIPs to the removed pod instance endpointIP.

Thanks for verifying! There could still be a conflict if someone is using 2 different UDP-based services with the same pods as endpoints right? If the same externalIP is used to connect to them, this code will clear valid connections for the existing service when the other one is being deleted.

**Comment by freehan** [comment]
It would be great this PR can also cover the IP from LoadbalancerStatus. Just need to mirror the logic for that case

**Comment by steven-sheehy** [comment]
Need to update PR title & description to reflect LB change. Also, is there a relevant ticket for LB @freehan ?

**Comment by prameshj** [comment]
> Need to update PR title & description to reflect LB change. Also, is there a relevant ticket for LB Minhan Xia ?
Can we reuse the same ticket for this too? We are just clearing conntrack entries for all IPs used to refer to the service, in addition to clusterIP

**Comment by steven-sheehy** [comment]
If you're asking me, I think it's fine to reuse ticket. I just wasn't sure if there was an existing issue for the LB IP that we can also link to.

**Review by freehan — APPROVED** [review]
/lgtm
/approve

**Comment by freehan** [comment]
> If you're asking me, I think it's fine to reuse ticket. I just wasn't sure if there was an existing issue for the LB IP that we can also link to.

I do not think there is an existing issue for LB IP. But I would imagine the same problem will hit UDP LBs.

**Review by thockin — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**